File: scripts/bulk_upload.py
import os
import glob
import cloudinary
import cloudinary.uploader
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore
from tinytag import TinyTag

# ==========================================
# CONFIGURATION
# ==========================================
# 1. Set the path to the folder containing your MP3 files
MP3_FOLDER_PATH = r"D:\Mp3 songs"

# 2. Set the path to your Firebase Admin SDK service account key JSON file
#    (Get this from Firebase Console -> Project Settings -> Service Accounts -> Generate new private key)
FIREBASE_SERVICE_ACCOUNT_KEY = r"C:\Users\psiva\Downloads\symphony-music-app-6eddc-firebase-adminsdk-fbsvc-8b2075f944.json"
# ==========================================

# Initialize Cloudinary
cloudinary.config(
    cloud_name="dx02qjcqn",
    api_key="424668629472834",
    api_secret="Ln3kEROFIDnMSimvM4n27EgavfI"
)

def init_firebase():
    """Initialize Firebase Admin SDK"""
    try:
        cred = credentials.Certificate(FIREBASE_SERVICE_ACCOUNT_KEY)
        firebase_admin.initialize_app(cred)
        return firestore.client()
    except Exception as e:
        print(f"Failed to initialize Firebase: {e}")
        return None

def process_and_upload_mp3s(directory_path, db):
    print(f"Starting bulk upload from: {directory_path}")
    print("-" * 50)
    
    # Use glob to find all .mp3 files recursively in the directory
    mp3_files = glob.glob(os.path.join(directory_path, '**/*.mp3'), recursive=True)
    
    if not mp3_files:
        print("No MP3 files found in the specified directory.")
        return

    print(f"Found {len(mp3_files)} MP3 files. Beginning processing...")
    
    success_count = 0
    failure_count = 0

    for file_path in mp3_files:
        filename = os.path.basename(file_path)
        print(f"\nProcessing: {filename}")
        
        try:
            # 1. Extract Metadata using tinytag
            tag = TinyTag.get(file_path)
            
            # Fallback to filename if title is missing in metadata
            title = tag.title
            if not title or title.strip() == "":
                title = os.path.splitext(filename)[0]
                
            artist = tag.artist
            if not artist or artist.strip() == "":
                artist = "Unknown Artist"
                
            print(f"  - Extracted metadata -> Title: '{title}', Artist: '{artist}'")

            # 2. Upload to Cloudinary
            print("  - Uploading audio to Cloudinary...")
            upload_result = cloudinary.uploader.upload(
                file_path, 
                resource_type="video", # Audio files MUST use resource_type 'video' in Cloudinary
                folder="symphony_audio" # Optional: keeps your Cloudinary dashboard clean
            )
            
            secure_url = upload_result.get('secure_url')
            if not secure_url:
                raise Exception("Cloudinary upload succeeded but no secure_url was returned.")
                
            print(f"  - Cloudinary upload successful. URL: {secure_url}")

            # 3. Save to Firestore
            print("  - Saving metadata to Firestore...")
            doc_ref = db.collection('songs').document()
            
            song_data = {
                'title': title.strip(),
                'artist': artist.strip(),
                'audioUrl': secure_url,
                'coverUrl': '', # Can be updated later by admin
                'isTrending': False,
                'createdAt': firestore.SERVER_TIMESTAMP
            }
            
            doc_ref.set(song_data)
            print(f"  - Successfully added to Firestore with ID: {doc_ref.id}")
            
            success_count += 1
            
        except Exception as e:
            print(f"  [ERROR] Failed to process {filename}: {e}")
            failure_count += 1
            continue # Important: continue to the next file if this one fails

    print("\n" + "=" * 50)
    print("Bulk Upload Completed!")
    print(f"Total Files Found: {len(mp3_files)}")
    print(f"Successfully Uploaded: {success_count}")
    print(f"Failed Uploads: {failure_count}")

if __name__ == "__main__":
    if MP3_FOLDER_PATH == "C:/path/to/your/mp3/folder":
        print("ERROR: Please open the script and update 'MP3_FOLDER_PATH' with your actual folder path.")
    elif FIREBASE_SERVICE_ACCOUNT_KEY == "C:/path/to/serviceAccountKey.json":
        print("ERROR: Please open the script and update 'FIREBASE_SERVICE_ACCOUNT_KEY' with the path to your Firebase credentials JSON file.")
    else:
        db = init_firebase()
        if db:
            process_and_upload_mp3s(MP3_FOLDER_PATH, db)

File: scripts/bulk_upload.js
const fs = require('fs');
const path = require('path');
const cloudinary = require('cloudinary').v2;
const admin = require('firebase-admin');

// ==========================================
// CONFIGURATION
// ==========================================
const MP3_FOLDER_PATH = "D:\\Mp3 songs";
const FIREBASE_SERVICE_ACCOUNT_KEY = "C:\\Users\\psiva\\Downloads\\symphony-music-app-6eddc-firebase-adminsdk-fbsvc-8b2075f944.json";

// Initialize Cloudinary
cloudinary.config({ 
  cloud_name: 'dx02qjcqn', 
  api_key: '424668629472834', 
  api_secret: 'Ln3kEROFIDnMSimvM4n27EgavfI' 
});

// Initialize Firebase
const serviceAccount = require(FIREBASE_SERVICE_ACCOUNT_KEY);
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});
const db = admin.firestore();

// Helpers
function getAllFiles(dirPath, arrayOfFiles) {
  const files = fs.readdirSync(dirPath);
  arrayOfFiles = arrayOfFiles || [];
  files.forEach(function(file) {
    if (fs.statSync(dirPath + "/" + file).isDirectory()) {
      arrayOfFiles = getAllFiles(dirPath + "/" + file, arrayOfFiles);
    } else {
      if (file.toLowerCase().endsWith('.mp3') || file.toLowerCase().endsWith('.m4a')) {
        arrayOfFiles.push(path.join(dirPath, file));
      }
    }
  });
  return arrayOfFiles;
}

async function uploadAudio(filePath) {
  return new Promise((resolve, reject) => {
    cloudinary.uploader.upload(filePath, { resource_type: "video", folder: "symphony_audio" }, function(error, result) {
      if (error) reject(error);
      else resolve(result.secure_url);
    });
  });
}

function toTitleCase(str) {
  if (!str) return "Unknown Artist";
  return str.split(' ').map(word => {
    if (word.length === 0) return word;
    return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
  }).join(' ');
}

function extractMetadata(rawTitle, rawArtist, filename) {
  let title = rawTitle || filename.replace(/\.(mp3|m4a)$/i, '');
  let artist = rawArtist || "Unknown Artist";

  // Clean common pirated site tags
  const badTagsRegex = /-?\s*(masstamilan|mp3bhai|masstelugu|friendstamilmp3|isaimini|starmusiq|sensongs).*$/i;
  
  title = title.replace(badTagsRegex, '').trim();
  artist = artist.replace(badTagsRegex, '').trim();
  
  if (artist === "" || artist.toLowerCase() === "unknown") {
    artist = "Unknown Artist";
  }

  // Check for Artist after hyphen
  const hyphenMatch = title.match(/^(.*?)\s*-\s*(.+)$/);
  if (hyphenMatch && artist === "Unknown Artist") {
    title = hyphenMatch[1];
    artist = hyphenMatch[2];
  }

  // Check for Artist in parentheses
  const parenMatch = title.match(/^(.*?)\s*\((.+?)\)$/);
  if (parenMatch && artist === "Unknown Artist") {
    title = parenMatch[1];
    artist = parenMatch[2];
  }

  title = title.replace(/\([^)]*\)/g, '').replace(/[- ]+$/, '').trim();
  artist = artist.replace(/\([^)]*\)/g, '').replace(/[- ]+$/, '').trim();

  return { 
    title: toTitleCase(title), 
    artist: toTitleCase(artist === "" ? "Unknown Artist" : artist) 
  };
}

async function searchITunes(title) {
  try {
    const searchQuery = title.split('-')[0].trim();
    const url = `https://itunes.apple.com/search?term=${encodeURIComponent(searchQuery)}&entity=song&limit=1`;
    const res = await fetch(url);
    const json = await res.json();
    if (json.results && json.results.length > 0) {
      const match = json.results[0];
      return {
        artist: match.artistName,
        coverUrl: match.artworkUrl100 ? match.artworkUrl100.replace('100x100bb', '600x600bb') : null
      };
    }
  } catch(e) {
    console.log("    -> iTunes lookup failed:", e.message);
  }
  return null;
}

async function uploadCoverArt(buffer) {
  return new Promise((resolve, reject) => {
    const stream = cloudinary.uploader.upload_stream(
      { folder: "symphony_covers" },
      function(error, result) {
        if (error) reject(error);
        else resolve(result.secure_url);
      }
    );
    stream.end(buffer);
  });
}

async function main() {
  console.log("Starting bulk upload with Cover Art, Smart Extraction & iTunes Fallback...");
  const files = getAllFiles(MP3_FOLDER_PATH);
  
  if (files.length === 0) {
      console.log("No audio files found in " + MP3_FOLDER_PATH);
      return;
  }
  
  console.log(`Found ${files.length} audio files. Uploading...`);
  const mm = await import('music-metadata');

  let successCount = 0;
  let updateCount = 0;
  let failureCount = 0;

  for (const file of files) {
    const filename = path.basename(file);
    console.log(`\nProcessing: ${filename}`);
    try {
      const metadata = await mm.parseFile(file);
      const extracted = extractMetadata(metadata.common.title, metadata.common.artist, filename);
      let title = extracted.title;
      let artist = extracted.artist;
      
      console.log(`  - Title: '${title}' | Artist: '${artist}'`);

      // Check if song exists first to save bandwidth and API calls
      const existing = await db.collection('songs').where('title', '==', title).limit(1).get();
      
      if (!existing.empty) {
        console.log(`  - Song '${title}' already exists. Skipping...`);
        // We skip updating because we already ran a cleanup script. 
        // If we want to add new songs, we just ignore existing ones.
        continue;
      }

      // Cover Art Extraction
      let coverUrl = '';
      if (metadata.common.picture && metadata.common.picture.length > 0) {
        console.log(`  - Found official MP3 cover art! Uploading...`);
        coverUrl = await uploadCoverArt(metadata.common.picture[0].data);
      } else {
        console.log(`  - No cover art found in file.`);
      }

      // iTunes Fallback for Artist / Cover Art
      if (artist === "Unknown Artist" || coverUrl === '') {
        console.log(`  - Missing artist or cover. Searching iTunes API...`);
        const itunesData = await searchITunes(title);
        if (itunesData) {
          if (artist === "Unknown Artist" && itunesData.artist) {
             artist = itunesData.artist;
             console.log(`    -> Found Artist on iTunes: ${artist}`);
          }
          if (coverUrl === '' && itunesData.coverUrl) {
             coverUrl = itunesData.coverUrl;
             console.log(`    -> Found Cover Art on iTunes!`);
          }
        } else {
          console.log(`    -> No iTunes match found.`);
        }
      }

      console.log(`  - Uploading audio to Cloudinary...`);
      const secure_url = await uploadAudio(file);

      console.log(`  - Saving to Firestore...`);
      await db.collection('songs').add({
        title: title,
        artist: artist,
        audioUrl: secure_url,
        coverUrl: coverUrl,
        isTrending: false,
        createdAt: admin.firestore.FieldValue.serverTimestamp()
      });
      console.log(`  - Success!`);
      successCount++;
    } catch (e) {
      console.error(`  [ERROR] Failed to process ${filename}:`, e.message);
      failureCount++;
    }
  }

  console.log("\n=========================================");
  console.log("Bulk Upload Completed!");
  console.log(`Total Files Found: ${files.length}`);
  console.log(`New Uploads: ${successCount}`);
  console.log(`Updated Existing: ${updateCount}`);
  console.log(`Failed: ${failureCount}`);
  
  process.exit(0);
}

main().catch(console.error);
